src/boot/efi/linux.c

```diff
@@ -4,89 +4,197 @@
 #include <efilib.h>
 
 #include "linux.h"
+#include "macro-fundamental.h"
+#include "missing_efi.h"
 #include "util.h"
 
-#ifdef __i386__
-#define __regparm0__ __attribute__((regparm(0)))
-#else
-#define __regparm0__
-#endif
+#define LINUX_INITRD_MEDIA_GUID \
+     {0x5568e427, 0x68fc, 0x4f3d, {0xac, 0x74, 0xca, 0x55, 0x52, 0x31, 0xcc, 0x68} }
+
+/* initrd that is served to the kernel */
+static VOID *_initrd_buffer = NULL;
+/* length of the initrd */
+static UINTN _initrd_length = 0;
+/* handle of the registerd protocol */
+static EFI_HANDLE _initrd_handle = NULL;
+
+/* work-arround for GCC */
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wmissing-braces"
+
+/* static structure for LINUX_INITRD_MEDIA device path
+   see https://github.com/torvalds/linux/blob/v5.13/drivers/firmware/efi/libstub/efi-stub-helper.c
+ */
+static const struct {
+        VENDOR_DEVICE_PATH vendor;
+        EFI_DEVICE_PATH end;
+} _packed_ efi_initrd_device_path = {
+        .vendor = {
+                {
+                        MEDIA_DEVICE_PATH,
+                        MEDIA_VENDOR_DP,
+                        sizeof(efi_initrd_device_path.vendor)
+                },
+                LINUX_INITRD_MEDIA_GUID
+        },
+        .end = {
+                END_DEVICE_PATH_TYPE,
+                END_ENTIRE_DEVICE_PATH_SUBTYPE,
+                sizeof(efi_initrd_device_path.end)
+        }
+};
+
+#pragma GCC diagnostic pop
+
+EFIAPI EFI_STATUS initrd_load_file(
+                EFI_LOAD_FILE_PROTOCOL *this,
+                EFI_DEVICE_PATH *file_path,
+                BOOLEAN boot_policy,
+                UINTN *buffer_size,
+                VOID *buffer
+);
+
+static const EFI_LOAD_FILE_INTERFACE load_file2_interface = {
+        .LoadFile = initrd_load_file
+};
+
+EFIAPI EFI_STATUS initrd_load_file(
+                EFI_LOAD_FILE_PROTOCOL* this,
+                EFI_DEVICE_PATH* file_path,
+                BOOLEAN boot_policy,
+                UINTN* buffer_size,
+                VOID* buffer) {
+
+        if (this != &load_file2_interface || !buffer_size || !file_path)
+                return EFI_INVALID_PARAMETER;
+        if (boot_policy)
+                return EFI_UNSUPPORTED;
+
+        if (!_initrd_length || !_initrd_buffer)
+                return EFI_NOT_FOUND;
+
+        if (!buffer || *buffer_size < _initrd_length) {
+                *buffer_size = _initrd_length;
+                return EFI_BUFFER_TOO_SMALL;
+        }
+
+        CopyMem(buffer, _initrd_buffer, _initrd_length);
+        *buffer_size = _initrd_length;
+        return EFI_SUCCESS;
+}
+
+static EFI_STATUS initrd_register(
+                VOID *initrd_buffer,
+                UINTN initrd_length) {
+
+        EFI_STATUS err;
+        EFI_DEVICE_PATH *dp = (EFI_DEVICE_PATH *) &efi_initrd_device_path;
+        EFI_HANDLE handle;
+
+        if (!initrd_buffer || !initrd_length)
+                return EFI_SUCCESS;
```

Comment:
doesn't matter, but maybe generate an error of the static vars are already initialized